PacoMojica/ondemand-go-bootcamp
# Final Deliverable

# What is in this PR?

Contains the requirements for the final deliverable, added a endpoint that reads the file concurrently. Also, changed stuff based on the comments on the previous PR

# How should this be manually tested?

Start the server by the default it should run on `http://localhost:8080/`, this can be changed in the file `config.yml`. The data of the app is stored in a CSV file called `db.csv`

The API has the following endpoints:

1. **`[GET]`** `/pokemon`: returns a list of all the pokemon in the DB
3. **`[GET]`** `/pokemon/<ID>`: returns the information of the pokemon that matches the ID,`null` otherwise
4. **`[POST]`** `/create-pokemon`: returns the data of the pokemon that was added to the DB
5. **`[GET]`** `/fetch-pokeapi`: queries the pokeAPI using a random pokemon ID, stores the response in the CSV and includes the pokemon in the response as a JSON
6. **`[GET]`** `/fetch-pokeapi/<ID>`: queries the pokeAPI with the provided pokemon ID, stores the response in the CSV and includes the pokemon in the response as a JSON
7. **`[GET]`** `/concurrent-pokemon?type=<odd|even>&items=<INT>&items_per_workers=<INT>`: as described in the requirements

> **`type`**: Only support "odd" or "even"
> **`items`**: Is an Int and is the amount of valid items you need to display as a response
> **`items_per_workers`**: Is an Int and is the amount of valid items the worker should append to the response

> The response should be displayed when:
> - The workers reached the limit
> - EOF
> - Valid items completed

### create pokemon body

```json
{
    "id": 132,
    "name": "ditto",
    "sprites": {
        "front_default": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/132.png"
    },
    "weight": 40,
    "height": 3,
    "base_experience": 101,
    "species": {
        "name": "ditto",
        "gender_rate": -1,
        "is_baby": false,
        "is_legendary": false,
        "is_mythical": false,
        "habitad": "urban"
    },
    "abilities": [
        {
            "ability": {
                "name": "limber",
                "url": "https://pokeapi.co/api/v2/ability/7/"
            },
            "is_hidden": false
        },
        {
            "ability": {
                "name": "imposter",
                "url": "https://pokeapi.co/api/v2/ability/150/"
            },
            "is_hidden": true
        }
    ],
    "moves": [
        {
            "move": {
                "name": "transform",
                "url": "https://pokeapi.co/api/v2/move/144/"
            }
        }
    ],
    "types": [
        {
            "type": {
                "name": "normal",
                "url": "https://pokeapi.co/api/v2/type/1/"
            }
        }
    ]
}
```
